File: solutions/cpp/Bignumber_multiplication.cc
// Copyright (c) 2015 Elements of Programming Interviews. All rights reserved.

#include <algorithm>
#include <cassert>
#include <iostream>
#include <iterator>
#include <random>
#include <sstream>
#include <string>
#include <vector>

#include "./execute-shell.h"

using std::cout;
using std::default_random_engine;
using std::endl;
using std::max;
using std::ostream_iterator;
using std::random_device;
using std::string;
using std::stringstream;
using std::uniform_int_distribution;
using std::vector;

// @include
vector<int> Multiply(vector<int> num1, vector<int> num2) {
  bool is_negative = (num1.front() < 0 && num2.front() >= 0) ||
                     (num1.front() >= 0 && num2.front() < 0);
  num1.front() = abs(num1.front()), num2.front() = abs(num2.front());

  // Reverses num1 and num2 to make multiplication easier.
  reverse(num1.begin(), num1.end());
  reverse(num2.begin(), num2.end());
  vector<int> result(num1.size() + num2.size(), 0);
  for (int i = 0; i < num1.size(); ++i) {
    for (int j = 0; j < num2.size(); ++j) {
      result[i + j] += num1[i] * num2[j];
      result[i + j + 1] += result[i + j] / 10;
      result[i + j] %= 10;
    }
  }

  // Skips the leading 0s and keeps one 0 if all are 0s.
  while (result.size() != 1 && result.back() == 0) {
    result.pop_back();
  }
  // Reverses result to get the most significant digit as the start of array.
  reverse(result.begin(), result.end());
  if (is_negative) {
    result.front() *= -1;
  }
  return result;
}
// @exclude

vector<int> RandVector(int len) {
  if (!len) return {0};

  default_random_engine gen((random_device())());
  uniform_int_distribution<int> dis(1, 9);
  vector<int> ret;
  ret.emplace_back(dis(gen));
  --len;
  while (len--) {
    uniform_int_distribution<int> dis(0, 9);
    ret.emplace_back(dis(gen));
  }

  uniform_int_distribution<int> positive_or_negative(0, 1);
  if (positive_or_negative(gen)) {
    ret.front() *= -1;
  }
  return ret;
}

bool EqualVector(const vector<int>& A, const vector<int>& B) {
  return A.size() == B.size() && equal(A.begin(), A.end(), B.begin());
}

void SimpleTest() {
  assert(EqualVector(Multiply({0}, {-1, 0, 0, 0}), {0}));
  assert(EqualVector(Multiply({0}, {1, 0, 0, 0}), {0}));
  assert(EqualVector(Multiply({9}, {9}), {8, 1}));
  assert(EqualVector(Multiply({9}, {9, 9, 9, 9}), {8, 9, 9, 9, 1}));
  assert(EqualVector(Multiply({1, 3, 1, 4, 1, 2}, {-1, 3, 1, 3, 3, 3, 2}),
                     {-1, 7, 2, 5, 8, 7, 5, 8, 4, 7, 8, 4}));
  assert(EqualVector(Multiply({7, 3}, {-3}), {-2, 1, 9}));
}

string VectorToString(const vector<int>& A) {
  stringstream converter;
  copy(A.begin(), A.end(), ostream_iterator<int>(converter, ""));
  return converter.str();
}

int main(int argc, char* argv[]) {
  SimpleTest();
  for (int times = 0; times < 1000; ++times) {
    default_random_engine gen((random_device())());
    uniform_int_distribution<int> dis(0, 19);
    vector<int> num1 = RandVector(dis(gen)), num2 = RandVector(dis(gen));
    auto res = Multiply(num1, num2);
    cout << VectorToString(num1) << " * " << VectorToString(num2) << " = "
         << VectorToString(res) << endl;
    string command = "bash -c 'bc <<<" + VectorToString(num1) + "*" +
                     VectorToString(num2) + "'";
    string result = execute_shell(command);
    cout << "answer = " << result;
    assert(result.substr(0, result.size() - 1) == VectorToString(res));
  }
  return 0;
}

File: solutions/python/bignumber_multiplication.py
import copy
import random
import sys


# @include
def multiply(num1, num2):
    sign = -1 if (num1[0] < 0) ^ (num2[0] < 0) else 1
    num1[0], num2[0] = abs(num1[0]), abs(num2[0])

    result = [0] * (len(num1) + len(num2))
    for i in reversed(range(len(num1))):
        for j in reversed(range(len(num2))):
            result[i + j + 1] += num1[i] * num2[j]
            result[i + j] += result[i + j + 1] // 10
            result[i + j + 1] %= 10

    # Remove the leading zeroes.
    result = result[next((i for i, x in enumerate(result) if x != 0),
                         len(result)):] or [0]
    return [sign * result[0]] + result[1:]
# @exclude


def rand_list(length):
    if length == 0:
        return [0]
    ret = [random.randint(1, 9)] + [random.randint(0, 9)
                                    for _ in range(length - 1)]
    if random.randint(0, 1) == 1:
        ret[0] *= -1
    return ret


def simple_test():
    assert multiply([0], [-1, 0, 0, 0]) == [0]
    assert multiply([0], [1, 0, 0, 0]) == [0]
    assert multiply([9], [9]) == [8, 1]
    assert multiply([9], [9, 9, 9, 9]) == [8, 9, 9, 9, 1]
    assert multiply(
        [1, 3, 1, 4, 1, 2],
        [-1, 3, 1, 3, 3, 3, 2]) == [-1, 7, 2, 5, 8, 7, 5, 8, 4, 7, 8, 4]
    assert multiply([7, 3], [-3]) == [-2, 1, 9]


def main():
    simple_test()
    for _ in range(1000):
        if len(sys.argv) == 3:
            s1, s2 = sys.argv[1], sys.argv[2]
        else:
            s1, s2 = rand_list(random.randint(0, 19)), rand_list(
                random.randint(0, 19))
        temp1, temp2 = copy.deepcopy(s1), copy.deepcopy(s2)
        res = multiply(temp1, temp2)
        print(s1, '*', s2, '=', res)
        result = int(''.join(map(str, s1))) * int(''.join(map(str, s2)))
        print('answer =', result)
        assert result == int(''.join(map(str, res)))


if __name__ == '__main__':
    main()
